src/format/OpVaultReader.cpp

```diff
@@ -0,0 +1,498 @@
+#include "OpVaultReader.h"
+#include "OpData01.h"
+
+#include "core/Group.h"
+#include "core/Tools.h"
+#include "crypto/CryptoHash.h"
+#include "crypto/SymmetricCipher.h"
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QUuid>
+#include <QtGui/QtGui>
+#include <gcrypt.h>
+
+static void populateCategoryGroups(Group* rootGroup);
+
+OpVaultReader::~OpVaultReader()
+{
+    delete m_masterKey;
+    delete m_masterHmacKey;
+    delete m_overviewKey;
+    delete m_overviewHmacKey;
+}
+
+Database* OpVaultReader::readDatabase(QDir& opdataDir, const void* passPhrase, size_t passPhraseLen)
+{
+    auto result = new Database();
+    readDatabase(opdataDir, result, passPhrase, passPhraseLen);
+    return result;
+}
+
+void OpVaultReader::readDatabase(QDir& opVaultDir, Database* db, const void* passPhrase, size_t passPhraseLen)
+{
+    if (!opVaultDir.exists()) {
+        m_error = true;
+        m_errorStr = "Directory .opvault must exist";
+        return;
+    }
+    if (!opVaultDir.isReadable()) {
+        m_error = true;
+        m_errorStr = "Directory .opvault must be readable";
+        return;
+    }
+
+    QDir defaultDir = QDir(opVaultDir);
+    if (!defaultDir.cd("default")) {
+        m_error = true;
+        m_errorStr = "Directory .opvault/default must exist";
+        return;
+    }
+    if (!defaultDir.isReadable()) {
+        m_error = true;
+        m_errorStr = "Directory .opvault/default must be readable";
+        return;
+    }
+
+    m_db = db;
+    m_meta = new Metadata(m_db);
+    auto rootGroup = new Group();
+    {
+        TimeInfo ti;
+        rootGroup->setTimeInfo(ti);
+        rootGroup->setUpdateTimeinfo(false);
+        rootGroup->setName("OPVault Root Group");
+        rootGroup->setUuid(QUuid::createUuid());
+    }
+    this->m_db->setRootGroup(rootGroup);
```

Comment:
Remove explicit `this`. I'd also prefer moving this directly below the allocation, so it's clear where this free-store object is parented.